llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp

```diff
@@ -948,6 +952,30 @@ bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
   return true;
 }
 
+bool RISCVInstructionSelector::selectIsFPClass(MachineInstr &MI,
+                                               MachineIRBuilder &MIB,
+                                               MachineRegisterInfo &MRI) const {
+  Register CheckResult = MI.getOperand(0).getReg();
+  Register Src = MI.getOperand(1).getReg();
+  int64_t MaskImm = MI.getOperand(2).getImm();
+  unsigned NewOpc = MRI.getType(Src).getSizeInBits() == 32 ? RISCV::FCLASS_S
+                                                           : RISCV::FCLASS_D;
+
+  Register FClassResult = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+  // Insert FCLASS_S/D.
+  auto FClass = MIB.buildInstr(NewOpc, {FClassResult}, {Src});
+  if (!FClass.constrainAllUses(TII, TRI, RBI))
+    return false;
+  // Insert AND to check Src aginst the mask.
+  auto And = MIB.buildInstr(RISCV::ANDI, {CheckResult}, {FClassResult})
+                 .addImm(MaskImm);
+  if (!And.constrainAllUses(TII, TRI, RBI))
+    return false;
+
```

Comment:
G_IS_FPCLASS is supposed to produce 0/1 in the LSB and have all other bits as 0. This code will can produce values between 0-1023.

Reply:
I think G_IS_FPCLASS's [doc](https://llvm.org/docs/GlobalISel/GenericOpcode.html#g-is-fpclass) states that any non-zero values is allowed to represent true: "...Returns non-zero (true) or zero (false). It’s target specific whether a true value is 1, ~0U, or some other non-zero value."

Reply:
This test case miscompiles

```
define zeroext i1 @foo(float %x) {                                               
  %a = call i1 @llvm.is.fpclass.f32(float %x, i32 1)                             
  ret i1 %a                                                                      
}                                                                                
                                                                                 
declare i1 @llvm.is.fpclass.f32(float, i32)  
```

The resulting code is
```
        fclass.s        a0, fa0
        andi    a0, a0, 256
        andi    a0, a0, 1
        ret
```

The second andi clears the bit from the first andi.

Reply:
I believe that the result should adhere to `TLI->getBooleanContents` which `ZeroOrOneBooleanContent`. That's the target specific part.

Reply:
You're right, I was using ANYEXT rather than ZEXT in my legalizer test and missed the ANDI it generated. 
It's fixed now.

Reply:
> I believe that the result should adhere to `TLI->getBooleanContents` which returns `ZeroOrOneBooleanContent` for RISC-V. That's the target specific part.

I didn't see this comment before pushing the latest update. Will fix in the next update

Reply:
I don't expect you to use `getBooleanContents`. It will always be ZeroOrOneBooleanContent for RISC-V. I'm fine with assuming that. If you were writing target independent code you would need to use it.